RecoMuon/SimMuonAlgos/plugins/MuonSimClassifier.cc

```diff
@@ -0,0 +1,574 @@
+    if (linkToGenParticles_) secondaries.reset(new reco::GenParticleCollection());
+
+    // loop on reco muons
+    for(size_t i = 0; i < nmu; ++i) {    
+        edm::RefToBase<reco::Muon> mu = muons->refAt(i);
+	edm::LogVerbatim("MuonSimClassifier") <<"\n reco::Muon # "<<i;
+
+        TrackingParticleRef        tp;
+        edm::RefToBase<reco::Muon> muMatchBack;
+        r2s_it match = recSimColl.find(mu);
+        s2r_it matchback;
+        if (match != recSimColl.end()) {
+            // match->second is vector, front is first element, first is the ref (second would be the quality)
+            tp = match->second.front().first;
+            tpId[i]          = tp.isNonnull() ? tp.key() : -1; // we check, even if null refs should not appear here at all
+            tpAssoQuality[i] = match->second.front().second;
+            s2r_it matchback = simRecColl.find(tp);
+            if (matchback != simRecColl.end()) {
+                muMatchBack = matchback->second.front().first;
+            } else {
+                edm::LogWarning("MuonSimClassifier") << "\n***WARNING:  This I do NOT understand: why no match back? *** \n";
+            }
+        } else if ((trackType_ == reco::GlobalTk) &&
+                    mu->isGlobalMuon()) {
+            // perform a second attempt, matching with the standalone muon
+            r2s_it matchSta = UpdSTA_recSimColl.find(mu);
+            if (matchSta != UpdSTA_recSimColl.end()) {
+                tp    = matchSta->second.front().first;
+                tpId[i]          = tp.isNonnull() ? tp.key() : -1; // we check, even if null refs should not appear here at all
+                tpAssoQuality[i] = matchSta->second.front().second;
+                s2r_it matchback = UpdSTA_simRecColl.find(tp);
+                if (matchback != UpdSTA_simRecColl.end()) {
+                    muMatchBack = matchback->second.front().first;
+                } else {
+                    edm::LogWarning("MuonSimClassifier") << "\n***WARNING:  This I do NOT understand: why no match back in UpdSTA? *** \n";
+                }
+            }
+        } else {
+            edm::LogVerbatim("MuonSimClassifier") <<"\t No matching TrackingParticle is found ";
+	}
+
+        if (tp.isNonnull()) {
+            bool isGhost = muMatchBack != mu;
+            if (isGhost) edm::LogVerbatim("MuonSimClassifier") <<"\t *** This seems a Duplicate muon ! classif[i] will be < 0 ***";
+
+	    // identify signal and pileup TP
+	    tpBx[i]      = tp->eventId().bunchCrossing();
+	    tpEv[i]      = tp->eventId().event();
+
+            hitsPdgId[i] = tp->pdgId();
+            prodRho[i]   = tp->vertex().Rho();
+            prodZ[i]     = tp->vertex().Z();
+
+	    // added info on GEANT process producing the TrackingParticle
+	    const std::vector<SimVertex> & G4Vs = tp->parentVertex()->g4Vertices();
+	    G4processType[i] = G4Vs[0].processType();
```

Comment:
start local variables and data members with lower case letter. Capitalization is reserved for class names.

Reply:
fixed